File: core/scripts/global.lua
function transition(state, event)
    if event.type == "auth.status" then
        return {
            state = event.error and "unhealthy" or "healthy",
            payload = {
                authenticated = event.authenticated,
                expires_at = event.expires_at,
                error = event.error,
                crashed = false
            }
        }
    end

    if event.type == "core.crash" then
        return {
            state = "crashed",
            payload = {
                authenticated = false,
                expires_at = nil,
                error = event.error,
                crashed = true
            }
        }
    end

    return {
        state = state,
        payload = {}
    }
end

File: core/scripts/global.ts
import type { PolicyEvent, PolicyResult } from "@inpainter/policy-runtime";

function luaTruthy(value: unknown): boolean {
  return value !== null && value !== undefined && value !== false;
}

export function transition(state: unknown, event: PolicyEvent): PolicyResult {
  if (event.type === "auth.status") {
    return {
      state: luaTruthy(event.error) ? "unhealthy" : "healthy",
      payload: {
        authenticated: event.authenticated ?? null,
        expires_at: event.expires_at ?? null,
        error: event.error ?? null,
        crashed: false,
      },
    };
  }
  if (event.type === "core.crash") {
    return {
      state: "crashed",
      payload: {
        authenticated: false,
        expires_at: null,
        error: event.error ?? null,
        crashed: true,
      },
    };
  }
  return { state, payload: {} };
}
